scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+                    index2(r, rs, j+1, j+1), rs[1], c, s) 
+
+        # update q
+        rot(m, col(q, qs, j), qs[0], col(q, qs, j+1), qs[0], c, s.conjugate())
+
+cdef void p_subdiag_qr(int m, int o, int n, blas_t* q, int* qs, blas_t* r, int* rs,
+                       int k, int p, blas_t* work) nogil:
+    """ Reduce a matrix r to upper triangular form by eliminating the lower p 
+        subdiagionals using reflectors. Both full and economic decompositions
+        are supported here.  In either case, q is (m,o) and r is (o,n)
+        
+        q and r must be fortran order here, with work at least max(m,n) long.
+    """
+    cdef int j
+    cdef int last
+    cdef blas_t tau
+    cdef blas_t rjj 
+    cdef int limit = min(m-1, n)
+    cdef char* sideR = 'R'
+    cdef char* sideL = 'L'
+
+    # R now has p subdiagonal values to be removed starting from col k.
+    for j in range(k, limit):
+        # length of the reflector
+        last = min(p+1, o-j)
+        rjj = index2(r, rs, j, j)[0]
+        larfg(last, &rjj, index2(r, rs, j+1, j), rs[0], &tau)
+        index2(r, rs, j, j)[0] = 1
+
+        # apply the reflector to r if necessary
+        if j+1 < n:
+            larf(sideL, last, n-j-1, index2(r, rs, j, j), rs[0],
+                    tau.conjugate(), index2(r, rs, j, j+1), rs[1], work)
+
+        # apply the reflector to q
+        larf(sideR, m, last, index2(r, rs, j, j), rs[0], tau,
+                index2(q, qs, 0, j), qs[1], work)
+
+        # rezero the householder vector we no longer need.
+        memset(index2(r, rs, j+1, j), 0, (last-1)*sizeof(blas_t))
+
+        # restore the rjj element
+        index2(r, rs, j, j)[0] = rjj
+
+def _reorth(cnp.ndarray q, cnp.ndarray u, rcond):
+    cdef cnp.ndarray s
+    cdef double rc_d
+    cdef float rc_f
+    cdef int m, n
+    cdef void* qp
+    cdef void* up
+    cdef void* sp
+    cdef int qs[2]
+    cdef int us[2]
+    cdef int ss[2]
+    cdef cnp.npy_intp size
+    cdef bint qisF
+
+    if q.ndim != 2: raise ValueError('q must be 2d')
+    m = q.shape[0]
+    n = q.shape[1]
+    if cnp.PyArray_CHKFLAGS(q, cnp.NPY_F_CONTIGUOUS):
+        qisF = True
+    elif cnp.PyArray_CHKFLAGS(q, cnp.NPY_C_CONTIGUOUS):
+        qisF = False
+    else:
+        raise ValueError('q must be one segment.')
+    if u.ndim != 1: raise ValueError('u must be 1d')
+    if u.shape[0] != m: raise ValueError('u.shape[0] must be q.shape[0]')
+    typecode = cnp.PyArray_TYPE(q)
+    if cnp.PyArray_TYPE(u) != typecode:
+        raise ValueError('q and u must have the same type.')
+
+    if not (typecode == cnp.NPY_FLOAT or typecode == cnp.NPY_DOUBLE \
+            or typecode == cnp.NPY_CFLOAT or typecode == cnp.NPY_CDOUBLE):
+        raise ValueError('q and u must be a blas compatible type: f d F or D')
+
+    q = validate_array(q, True)
+    u = validate_array(u, True)
+
+    qp = extract(q, qs)
+    up = extract(u, us)
+
+    size = 2*n
+    s = cnp.PyArray_ZEROS(1, &size, typecode, 1)
```

Comment:
It returns a python object, I think cython handles checking the output for me.  I'll look at the generate code to double check, but I think this okay.